### DISCUSSION TITLE
Prompt Composability

### DISCUSSION THREAD

**Discussion by meissnereric** [langfuse/langfuse#1660] [Ideas]
### Describe the feature or potential improvement

 There are certain things I need to add to all or most of my system's prompts and we compose them just before .invoke right now (using Langchain), but would be great if I could do that in your system and get it all stored there. 
 
 For instance, we usually load in a subset of previous user conversation and that template would be helpful to add in as a component in your PromptManagement tool. 
 
 Potentially then when I load prompt C (that is composed of A + B for instance) it would automatically load both of those sub-prompts and use the config from prompt C.

**Comment by marcklingen** [comment]
Great idea, I imagine this to work like this (with some @mention UI):

Prompt A
```
prompt
```

Prompt B
```
prompt
```

Prompt C
```
@[Prompt A] @[PromptB]
```

**Comment by zoltan-fedor** [comment]
Funny, I just wanted to raise the same idea.

We have a ReAct agent, which uses multiple tools (tools are dynamically selected), so we have one prompt for the ReAct agent itself and one for each tool. At runtime we put those together (ReAct prompt + multiple tool prompts) and supply that to the LangChain ReAct agent module.

This works, but now the generations in Langfuse are not aware which prompt was used - because in fact there were multiple.

So at minimum it would be nice to have an option to tell Langfuse that this call is based on multiple prompts (which were used to compose the actual prompt).

Ps.: Although this is not exactly the same as above, as we are constructing the final prompt from the tool prompt dynamically at runtime, so the following would not be sufficient:
```@[Prompt A] @[PromptB]```

**Reply by marcklingen** [reply]
Now tracked independently here, thanks for sharing!

https://github.com/orgs/langfuse/discussions/4040

**Comment by jasonslyvia** [comment]
We now have a bunch of different prompts share the same `system` prompt, for now we just copy & paste, it would be nice to compose them.

**Comment by isc-psulin** [comment]
The current implementation is a simple template, supporting {{variable}} interpolation. Allow a template to be a variable and you are halfway there or just maybe use an existing template library? Composability is required unless everyone is OK with copy/paste and find/replace.

**Comment by marcklingen** [comment]
what makes this a bit more complicated

- server side caching that needs to be invalidated for all prompts that are related to the prompt that was updated. simple solution, invalidate all prompts of the project as reads are cheap
- render aggregated prompt and prompt variables nicely

**Comment by sfarthin** [comment]
I would assume that all composed prompts would be versioned. It would indeed be unexpected if a change to Prompt A or B dynamically altered Prompt C, as this could significantly affect the output and potentially disrupt specific downstream prompts in production.

Users could still define a prompt simply as:

```
@[Prompt A] @[Prompt B]
```

This configuration would function as a snapshot of the prompts at that time, effectively referencing specific versions like:

```
@[Prompt A:22] @[Prompt B:34]
```

To accommodate additional changes, an explicit version update would be required. This concept mirrors how Continuous Integration (CI) systems process dependabot updates, ensuring that each change is explicitly tested and validated before being integrated into the production environment. This approach ensures stability and consistency, allowing for controlled updates while maintaining reliability.

I do worry this could lead to an issue is similar to the [fragile base class problem](https://en.wikipedia.org/wiki/Fragile_base_class), especially within deeply nested hierarchies.

**Reply by marcklingen** [reply]
I agree that this could lead to a cascading effect of updates. While it's straightforward to visualize in the UI at one level, it becomes more complex with multiple levels of different permissions.

Referencing a version of another prompt and providing a UI to identify which prompts depend on a specific prompt would be helpful and predictable.

Additionally, some teams desire the composability of prompts to simplify the updating of standard components. However, if each update of dependent prompts requires user interaction or approval, it diminishes the value of prompt composition.

**Comment by simone3991** [comment]
Same need here, but for different reasons - thought you could use another angle. 

We serve different enterprises with the same solution. Part of the prompting is fixed, each version update representing an improvement for all. But the remaining part is account-specific, one per workspace. As of now the merging is done by code, but composing prompt would be a killer feature to lower our operations.

**Comment by SanjayBysani** [comment]
Hi, we're looking for this functionality, and I stumbled across this in the [_Planned_](https://langfuse.com/docs/roadmap#-planned) section. Curious, is there some expected time period for features called out in the planned section ?

**Reply by marcklingen** [reply]
Generally we do not communicate strict timelines as priorities can change if urgent developments require new features.

Regarding prompt composability, @hassiebp is already working on a spec and I'd expect this to land over the next days/weeks (rather soonish)

**Comment by hassiebp** [comment]
We have just shipped suppport for prompt composability! 🫡

See our [changelog](https://langfuse.com/changelog/2025-03-12-prompt-composability) for more details

**Reply by SanjayBysani** [reply]
Amazing, thanks!

**Comment by emilrueh** [comment]
Hi, I am looking to use prompt composability but can not see any option to do so from code.
Neither of your doc pages talks about how to pass a prompt reference from code.
Without that I do not really see the use for it beyond basic structuring in the UI.

I'd need the option to pass it into a prompt when being fetched from the python client sdk.

Either via a simple placeholder that we can pass the prompt reference string into and that gets parsed into the prompt after having been inserted into the prompt via placeholder - or via some additional feature outside of placeholders.

It seems that currently running prompt.compile() will not correctly parse the prompt reference if passed via placeholders as probably the placeholders are being parsed after looking for a referenced prompt. Perhaps turning around this order would allow for that feature?

**Reply by hassiebp** [reply]
> I'd need the option to pass it into a prompt when being fetched from the python client sdk.

This is currently not supported. Prompt references cannot be passed dynamically, but must be declared at prompt creation and be part of the prompt.

**Reply by emilrueh** [reply]
I see. I believe this feature would really open up langfuse prompt management to be truly modular though. What do you think?
